### PR TITLE
2015 04 jk add kiezkassen process type

### PR DESCRIPTION
**Pull request by joka**
Kiezkassen process workflow is still missing, due to lack of specification.

### CODE REVIEW

**Comment by MagdaN** [comment]
@joka: I added the script - still needs gardening. To run it you need to install ogr2ogr (GDAL). Meine Version ist GDAL 1.10.1.

File: src/adhocracy_core/adhocracy_core/sheets/geo.py [hunk 1]
```diff
@@ -32,23 +36,173 @@ class WebMercatorLatitude(colander.SchemaNode):
...
     `y`: latitude in web mercator
     """
+    default = (0, 0)
+    missing = colander.drop
 
     x = WebMercatorLongitude()
     y = WebMercatorLatitude()
 
 
+class LineString(colander.SequenceSchema):
```
Comment (pallix) [inline src/adhocracy_core/adhocracy_core/sheets/geo.py]: Not sure I understand why the class is named so.

File: src/adhocracy_core/adhocracy_core/sheets/test_geo.py [hunk 2]
```diff
@@ -9,42 +103,84 @@ def inst(self):
...
-        assert inst['x'].validator.min == -20026376.3
-        assert inst['y'].validator.max == 20048966.1
-        assert inst['y'].validator.min == -20048966.1
+    def test_deserialize_valid(self, inst):
+        cstruct = {'coordinates': ('0', '0')}
+        assert inst.deserialize(cstruct) == {'coordinates': (0, 0)}
+
+    @mark.parametrize('x,y', [('20026376.4', '0'),
+                              ('-20026377', '0'),
+                              # actually we want -20026376.4 but its not working
```
Comment (pallix) [inline src/adhocracy_core/adhocracy_core/sheets/test_geo.py]: what is not working with this value? I had the test working with this value.

File: src/adhocracy_core/adhocracy_core/sheets/test_description.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        assert inst.meta.schema_class == DescriptionSchema
+        assert inst.meta.editable is True
+        assert inst.meta.create_mandatory is False
+
+    def test_get_empty(self, meta, context):
+        inst = meta.sheet_class(meta, context)
+        assert inst.get() == {'description': ''}
+
+
+def test_includeme_register_description_sheet(config):
```
Comment (pallix) [inline src/adhocracy_core/adhocracy_core/sheets/test_description.py]: Maybe we should move this test in the class, like here: https://github.com/liqd/adhocracy3.mercator/pull/1034/files#diff-81b4a36673cf425b0f8f52cc4daf697eR96

File: src/adhocracy_core/adhocracy_core/evolution/__init__.py [hunk 4]
```diff
@@ -1,13 +1,104 @@
...
+    """Add permission to use the kiezkassen process."""
+    logger.info('Running evolve step:' + add_kiezkassen_permissions.__doc__)
+
+    registry = get_current_registry()
+    acl = get_acl(root)
+    new_acl = [(Allow, 'role:contributor', 'add_kiezkassen_proposal'),
+               (Allow, 'role:creator', 'add_kiezkassen_proposal_version'),
+               (Allow, 'role:admin', 'add_kiezkassen_process'),
+               (Allow, 'role:admin', 'add_process')]
+    updated_acl = acl + new_acl
```
Comment (pallix) [inline src/adhocracy_core/adhocracy_core/evolution/__init__.py]: new_acl should be first

File: src/adhocracy_core/adhocracy_core/evolution/test_init.py [hunk 5]
```diff
@@ -36,3 +47,84 @@ def tearDown(self):
 
     def test_includeme_add_directives(self):
         assert 'add_evolution_step' in self.config.registry._directives
+
+from pytest import fixture
```
Comment (pallix) [inline src/adhocracy_core/adhocracy_core/evolution/test_init.py]: why not do the import on top?

**Comment by nidico** [comment]
Frontend code and backend geo structure modelling looks good to me.

File: src/adhocracy_core/adhocracy_core/evolution/__init__.py [hunk 6]
```diff
@@ -1,13 +1,104 @@
...
+
+def evolve1_add_title_sheet_to_pools(root: IPool):  # pragma: no cover
+    """Add title sheet to basic pools and asset pools."""
+    migrate_new_sheet(root, IBasicPool, ITitle, IPool,
+                      remove_isheet_old=False)
+    migrate_new_sheet(root, IPoolWithAssets, ITitle, IPool,
+                      remove_isheet_old=False)
+
+
+def add_kiezkassen_permissions(root):
```
Comment (pallix) [inline src/adhocracy_core/adhocracy_core/evolution/__init__.py]: should this be moved to the meinberlin package?

File: src/adhocracy_core/adhocracy_core/sheets/test_title.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        assert inst.meta.schema_class == TitleSchema
+        assert inst.meta.editable is True
+        assert inst.meta.create_mandatory is False
+
+    def test_get_empty(self, meta, context):
+        inst = meta.sheet_class(meta, context)
+        assert inst.get() == {'title': ''}
+
+
+def test_includeme_register_title_sheet(config):
```
Comment (pallix) [inline src/adhocracy_core/adhocracy_core/sheets/test_title.py]: same as above. Why not part of the class?

**Comment by pallix** [comment]
Backend code looks very good to me.

**Comment by MagdaN** [comment]
@pallix @joka I updated the import script, please check if it is allright for you. To run this, make sure GDAL (*Version 1.10.1*) is installed. 

First run:  bin/import_bezirke ./etc/development.ini
Second run:  bin/import_bezirksregions ./etc/development.ini

You should be able to see all created MultiPoligons under: http://localhost:6541/locations/ (you might need to delete the browser chache)

**Pull request merged**